lib/spack/spack/cmd/buildcache.py

```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+    name: Optional[str] = None,
+    update_keys: bool = False,
+):
+    """update a buildcache view index"""
+    # OCI images do not support views.
+    try:
+        spack.oci.oci.image_from_mirror(mirror)
+        raise spack.error.SpackError("OCI build caches do not support index views")
+    except ValueError:
+        pass
+
+    if update_mode == ViewUpdateMode.APPEND:
+        tty.warn(
+            "Appending to a view index does not guarantee idempotent write when contending "
+            "with multiple writers. This feature is meant to be used by a single process."
+        )
+
+    # Otherwise, assume a normal mirror.
+    url = mirror.push_url
+
+    if (name and mirror.push_view) and not name == mirror.push_view:
+        tty.warn(
+            (
+                f"Updating index view with name ({name}), which is different than "
+                f"the configured name ({mirror.push_view}) for the mirror {mirror.name}"
+            )
+        )
+
+    name = name or mirror.push_view
+    assert name
+
+    url_and_version = spack.binary_distribution.MirrorMetadata(
+        url, spack.binary_distribution.CURRENT_BUILD_CACHE_LAYOUT_VERSION, name
+    )
+
+    # Check if the index already exists, if it does make sure there is a copy in the
+    # local cache.
+    index_exists = True
+    try:
+        BINARY_INDEX._fetch_and_cache_index(url_and_version)
+    except spack.binary_distribution.BuildcacheIndexNotExists:
+        index_exists = False
+
+    if index_exists and update_mode == ViewUpdateMode.CREATE:
+        raise spack.error.SpackError(
+            "Index already exists. To overwrite or update pass --force or --append respectively"
+        )
+
+    hashes = []
+    if sources:
+        for source in sources:
+            tty.debug(f"reading specs from source: {source}")
+            hashes.extend(read_concrete_hashes(source))
+    else:
+        # Get hashes in the current active environment
+        env = spack.cmd.require_active_env(cmd_name="buildcache update-view")
+        hashes = env.all_hashes()
+
+    if not hashes:
+        tty.warn("No specs found for view, creating an empty index")
+
+    filter_fn = lambda x: x in hashes
+
+    with tempfile.TemporaryDirectory(dir=spack.stage.get_stage_root()) as tmpdir:
+        # Initialize a database
+        db = spack.binary_distribution.BuildCacheDatabase(tmpdir)
+        db._write()
+
+        if update_mode == ViewUpdateMode.APPEND:
+            # Load the current state of the view index from the cache into the database
+            cache_index = BINARY_INDEX._local_index_cache.get(str(url_and_version))
+            if cache_index:
+                cache_key = cache_index["index_path"]
+                db._read_from_file(BINARY_INDEX._index_file_cache.cache_path(cache_key))
+
+        spack.binary_distribution._url_generate_package_index(url, tmpdir, db, name, filter_fn)
+
+    if update_keys:
+        mirror_update_keys(mirror)
+
+
+def check_index_fn(args):
```

Comment:
I think it would make sense to push most of the content of this method into the `spack.mirror` module and leave the output and argument handling here. That way tests and other internal code can use this check as well.

But I'm ok with leaving that as TODO for a next PR to get this in.

Reply:
Thanks. Yeah I think the relationship between `cmd.mirror` and `cmd.buildcache` needs work in general. It is far too confusing.

Reply:
Thanks. Yeah I think the relationship between `cmd.mirror` and `cmd.buildcache` needs work in general. It is far too confusing.